File: src/julia/kamada_kawai/algo/algo.jl
module Algo

include("./mem_seq.jl")
using .Mem_seq
include("./mem_par.jl")
using .Mem_par
include("./par.jl")
using .Par

using BenchmarkTools

using ..Kamada_Kawai2csv
using ..Kamada_Kawai

include("../../common/write.jl")
using .Write
include("../../common/file.jl")
using .File

export algo

function output_algo(kk::KamadaKawai, original::Vector{Coord}, elapsed, name::String, test_id::String)
	# --- WRITE RESULTS TO FILE --- #
	algoresult = generateFilename(name, test_id, "csv")
	algoresult_fp = open(algoresult, "w")
	writeState(algoresult_fp, original)
	writeState(algoresult_fp, kk.coords)
	close(algoresult_fp)
	# --- WRITE TIME TO FILE --- #
	algotime = generateFilename(name, test_id, "txt")
	writeTime(algotime, elapsed)
end

function run_algo(kk::KamadaKawai, f::Function, name::String, test_id::String, repetitions::Int)
	# --- SETUP --- #
	original = set_original_coords(kk)

	# Warmup
	get_original_coords(kk, original)

    # --- BENCHMARK --- #
	# f(kk)
	# bench = @benchmark begin
	#     get_original_coords($kk, $original)
	#     $f($kk)
	# end samples=repetitions evals=1
	# avg_time = median(bench).time / 1e9
	
	# bench = @benchmark $f($kk) setup=(get_original_coords($kk, $original)) samples=repetitions evals=1
	# avg_time = median(bench).time / 1e9
	
	times = Float64[]
    for i in 1:repetitions
		get_original_coords(kk, original)
        t = @elapsed result = f(kk)
        push!(times, t)
    end
    avg_time = median(times)

	# --- OUTPUT --- #
	output_algo(kk, original, avg_time, name, test_id)

	# --- RESET DATA --- #
	get_original_coords(kk, original)
end

function algo(kk::KamadaKawai, test_id::String, repetitions::Int)
	# run_algo(kk, Mem_seq.mem_seq, "mem_seq", test_id, repetitions)
	run_algo(kk, Mem_par.mem_par, "mem_par", test_id, repetitions)
	# run_algo(kk, Par.par, "par", test_id, repetitions)

end # algo

end # module

File: src/go/kamada_kawai/algo/algo.go
package algo

import (
	parallelNP "golang/common"
	"os"
	"time"
)

func output_algo(kk *KamadaKawai, original []Coord, elapsed time.Duration, name string, test_id string) {
	// --- WRITE RESULTS TO FILE --- //
	algoresult := parallelNP.GenerateFilename(name, test_id, "csv")
	algoresult_fp, err := os.Create(algoresult)
	if err != nil {
		panic(err)
	}
	WriteState(algoresult_fp, original)
	WriteState(algoresult_fp, kk.Coords)
	defer algoresult_fp.Close()
	if err != nil {
		parallelNP.IOError("main.go", "Error closing the file", err)
	}

	// --- WRITE TIME TO FILE --- //
	algotime := parallelNP.GenerateFilename(name, test_id, "txt")
	algotime_fp, err := os.Create(algotime)
	if err != nil {
		parallelNP.IOError("main.go", "Error creating the file", err)
		return
	}
	defer algotime_fp.Close()
	parallelNP.WriteTime(algotime_fp, elapsed)
}

func run_algo(kk *KamadaKawai, algoFunc func(), name string, test_id string, repetitions int) {
	// --- SETUP --- //
	original := kk.set_original_coords()

	// --- RUN ALGORITHM R times --- //
	var avg_time time.Duration = 0.0
	for i := 0; i < repetitions; i++ {
		// --- RESET DATA --- //
		kk.get_original_coords(original)

		// --- EXECUTION --- //
		start := time.Now()
		algoFunc()
		elapsed := time.Since(start)
		avg_time = avg_time + elapsed
	}
	avg_time = avg_time / time.Duration(repetitions)

	// --- OUTPUT --- //
	output_algo(kk, original, avg_time, name, test_id)

	// --- RESET DATA --- //
	kk.get_original_coords(original)
}

func (kk *KamadaKawai) Algo(test_id string, repetitions int) error {
	run_algo(kk, kk.Sgl_seq, "sgl_seq", test_id, repetitions)
	// run_algo(kk, kk.Sgl_par, "sgl_par", test_id, repetitions)

	return nil
}
